minimapstyle.cpp

```diff
@@ -444,20 +446,51 @@ class MinimapStyleObject : public QObject
     void updateSubControlRects()
     {
         QScrollBar *scrollbar = m_editor->verticalScrollBar();
-        int viewPortLineCount = qRound(m_factor * (m_lineCount - scrollbar->maximum()));
+
+        int viewportHeight = m_editor->viewport()->height();
+        int lineHeight = m_editor->fontMetrics().lineSpacing();
+        int actualLinesPerPage = qMax(1, viewportHeight / lineHeight);
+
+        int viewPortLineCount = qRound(m_factor * actualLinesPerPage);
+        viewPortLineCount = qMax(1, qMin(viewPortLineCount, m_groove.height()));
+
         int w = scrollbar->width();
         int h = scrollbar->height();
         int value = scrollbar->value();
-        int realValue = qRound(m_factor * value);
         int min = scrollbar->minimum();
         int max = scrollbar->maximum();
-        m_addPage = value < max ? QRect(0,
-                                        realValue + viewPortLineCount,
-                                        w,
-                                        h - viewPortLineCount - realValue)
-                                : QRect();
-        m_subPage = value > min ? QRect(0, 0, w, realValue) : QRect();
+
+        int actualContentHeight;
+        if (m_factor < 1.0) {
+            // When zoomed out, content is scaled to fit
+            actualContentHeight = qRound(m_lineCount * m_factor);
+        } else {
+            // When 1:1 or zoomed in, each line takes 1 pixel
+            actualContentHeight = qMin(m_lineCount, h);
+        }
+        actualContentHeight = qMax(1, actualContentHeight);
+
+        int realValue = 0;
+        if (max > min && actualContentHeight > viewPortLineCount) {
+            qreal scrollRatio = static_cast<qreal>(value - min) / (max - min);
+            int maxSliderTop = actualContentHeight - viewPortLineCount;
+            realValue = qRound(scrollRatio * maxSliderTop);
+            realValue = qMax(0, qMin(realValue, maxSliderTop));
+        }
+
+        if (realValue + viewPortLineCount > actualContentHeight) {
+            realValue = actualContentHeight - viewPortLineCount;
+        }
+        realValue = qMax(0, realValue);
+
+        m_addPage = (realValue + viewPortLineCount < h) ?
+        QRect(0, realValue + viewPortLineCount, w, h - realValue - viewPortLineCount) :
+        QRect();
+        m_subPage = (realValue > 0) ?
+        QRect(0, 0, w, realValue) :
+        QRect();
```

Comment:
Can we format these so its more obvious that these are ternary operator statements? :)